src/main/java/org/broadinstitute/hellbender/engine/ReferenceDataSourceFromReferenceBases.java

```diff
@@ -0,0 +1,128 @@
+    public ReferenceDataSourceFromReferenceBases(final ReferenceBases bases, final SAMSequenceDictionary referenceSequenceDictionary) {
+        this.bases = Utils.nonNull(bases);
+        this.sequenceDictionary = referenceSequenceDictionary;
+    }
+
+    /**
+     * Start an iteration over the entire reference. Not yet supported!
+     *
+     * See the BaseUtils class for guidance on how to work with bases in this format.
+     *
+     * @return iterator over all bases in this reference
+     */
+    @Override
+    public Iterator<Byte> iterator() {
+        throw new UnsupportedOperationException("Iteration over entire reference not yet implemented");
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back an iterator over the bases spanning that interval.
+     *
+     * See the BaseUtils class for guidance on how to work with bases in this format.
+     *
+     * @param interval query interval
+     * @return iterator over the bases spanning the query interval
+     */
+    @Override
+    public Iterator<Byte> query( final SimpleInterval interval ) {
+
+        int startIndex = (bases.getInterval().getStart() - interval.getStart());
+        // compute how many bases we're reading (+1 because those are closed intervals),
+        // then add however many bases we're skipping at the front.
+        int stopIndex = (interval.getStart() - interval.getEnd() + 1) + startIndex;
+
+        return new ByteArrayIterator(bases.getBases(), startIndex, stopIndex);
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back all bases spanning that interval at once.
+     * Call getBases() on the returned ReferenceSequence to get the actual reference bases. See the BaseUtils
+     * class for guidance on how to work with bases in this format.
+     *
+     * @param interval query interval
+     * @return a ReferenceSequence containing all bases spanning the query interval, prefetched
+     */
+    @Override
+    public ReferenceSequence queryAndPrefetch( final SimpleInterval interval ) {
+        return queryAndPrefetch(interval.getContig(), interval.getStart(), interval.getEnd());
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back all bases spanning that interval at once.
+     * Call getBases() on the returned ReferenceSequence to get the actual reference bases. See the BaseUtils
+     * class for guidance on how to work with bases in this format.
+     *
+     * @param contig query interval contig
+     * @param start query interval start
+     * @param stop query interval stop (included)
+     * @return a ReferenceSequence containing all bases spanning the query interval, prefetched
+     */
+    @Override
+    public ReferenceSequence queryAndPrefetch( final String contig, final long start , final long stop) {
+        final int contigIndex = sequenceDictionary.getSequenceIndex(contig);
+        int startIndex = (int)(bases.getInterval().getStart() - start);
```

Comment:
I think you didn't trigger this bug because you are querying in the BAQ code in `BaseRecalibratorFn.calculateBAQArray()` for EXACTLY the reference bases you have stored, and when you go to get the subset of bases overlapping each read in `BaseRecalibratorFn.calculateIsSNP()` you bypass `queryAndPrefetch()/query() entirely.